packages/core/src/Serialization/TransactionWitnessSet/TransactionWitnessSet.ts

```diff
@@ -206,18 +247,44 @@ export class TransactionWitnessSet {
 
           reader.readEndArray();
           break;
-        case 5n:
+        case 5n: {
+          const redeemers: Redeemer[] = [];
           witness.setRedeemers(new Array<Redeemer>());
-          reader.readStartArray();
-
-          while (reader.peekState() !== CborReaderState.EndArray) {
-            witness.redeemers()!.push(Redeemer.fromCbor(HexBlob.fromBytes(reader.readEncodedValue())));
+          if (reader.peekState() === CborReaderState.StartMap) {
+            // { + [ tag: redeemer_tag, index: uint ] => [ data: plutus_data, ex_units: ex_units ] }
+            reader.readStartMap();
+            while (reader.peekState() !== CborReaderState.EndMap) {
+              // Read key cbor
+              reader.readStartArray();
```

Comment:
We can assert here that the array has two elements, and throw otherwise as invalid CBOR for the redeemer key (same with the value)